Fix #2990: Implement workaround to make MathExpressionInput usable on mobile devices.

Oishikatta opened the pull request:
This implements a workaround to make the MathExpressionInput usable on mobile devices.

Uses a transparent (opacity: 0) button element over the Guppy div to call the focus() method on a transparent/1px text input during the onclick event, in order to display the keyboard on mobile devices.

[The test used for mobile devices is matching the string "Mobi" in the User-Agent as recommended by MDN.
](https://developer.mozilla.org/en-US/docs/Web/HTTP/Browser_detection_using_the_user_agent) Larger touchscreen-only devices, e.g. Windows tablets, would not be matched by this test but typically have a button for the user to bring up the keyboard themselves.

Issues with this method include that the user cannot select text and cannot use arrow keys (deletion/editing only by backspace). These are on mobile devices only, behavior in non-mobile browsers is unchanged.


Status by device type:

| Device | Browser | Status |
| --- | --- | --- |
| Android 4.x,6.x,7.x | Android Browser, Chrome | No other issues observed.
| Windows Phone 8.1 | IE Mobile 11 | Keyboard appears to work, but there may be upstream issues with IE 11 Mobile compatibility. Basic expressions only (no exponents, etc).
| iPhone | Safari (iOS 10.2) | Keyboard can only be opened once per MathExpressionInput element. Not a huge issue since a new MathExpressionInput element is created each time an incorrect answer is submitted.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.html:
```diff
@@ -5,6 +5,11 @@
   <div class="guppy-div"
        style="width: 100%; height: 100px; padding: 5px;">
   </div>
+  <button ng-if="isMobileOverlayShown" id="startMathInputButton" style="opacity: 0; position: absolute; z-index: 2;"></button>
```
Turn these long styles into classes.

seanlip replied:
Also, if it's a variable, let's go with $scope.mobileOverlayIsShown.

Oishikatta replied:
Done.

Oishikatta replied:
Done.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+              var guppyOffset = $(guppyDivElt).position();
+              var guppySize = document.querySelector('.guppy-div')
+                .getBoundingClientRect();
+              console.log(guppySize, $('#startMathInputButton'));
```
Drop console.logs before pushing.

Oishikatta replied:
Done.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+             */
+            var positionButtonOverlay = function() {
+              var guppyOffset = $(guppyDivElt).position();
+              var guppySize = document.querySelector('.guppy-div')
```
Don't we already have the guppyDivElt?

Oishikatta replied:
We did, fixed.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+
+            // The focus() call must be in a click event handler and on a text
+            // field to make the mobile keyboard appear.
+            $('#startMathInputButton').click(function() {
```
Can this be done by adding ng-click to the element (i.e., the Angular way)?

Oishikatta replied:
Done.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+              guppyInstance.activate();
+
+              var fakeInputElement = document.
+                querySelector('#fakeInputForMathExpression');
```
Prefer breaking after the '(' rather than breaking at the '.'.

Oishikatta replied:
Done.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+
+              // Place the cursor at the end of the text input, so that the
+              // user can use backspace to delete.
+              fakeInputElement.setSelectionRange(fakeInputElement.value.length,
```
Break after '('.

Oishikatta replied:
Done.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+                fakeInputElement.value.length);
+            });
+
+            // Mapping characters back to Mousetrap codes.
```
The constant name and the comments are all a bit cryptic to me. Can you explain more?

Oishikatta replied:
Guppy uses the Mousetrap [(ccampbell/mousetrap)](https://github.com/ccampbell/mousetrap) library for keyboard input. Mousetrap has a function call which allows sending key events, but for symbols like "^" they need to be sent as "shift+6" to be recognized. This object is used to convert symbols into strings that Mousetrap accepts via the Mousetrap.trigger() function.

Would that be a better comment? I'm not sure how it should be explained.

seanlip replied:
Yes, it would be a better comment! What you wrote here (in the github comment thread) is perfect.

Oishikatta replied:
Done.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
                 GLOBALS.ASSET_DIR_PREFIX +
                 '/assets/overrides/guppy/oppia_symbols.json');
+
+              if (/Mobi/.test(navigator.userAgent)) {
```
Break this test out into a separate function -- perhaps put it in deviceInfoService in app.js.

Oishikatta replied:
Added to/using deviceInfoService.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+
+              if (/Mobi/.test(navigator.userAgent)) {
+                $scope.isMobileOverlayShown = true;
+                // Wait for the scope change to apply.
```
Why? (Comment unclear by itself; maybe elaborate more.)

Oishikatta replied:
Added elaboration to comment.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+              'input change compositionupdate keydown',
+              oppiaDebouncer.debounce(function() {
+                setGuppyContentFromInput();
+              }, 100))
```
Move last closing paren to next line and follow it immediately with the second .on. E.g.:

```
$(...).on(
  'input ...',
  oppiaDebouncer.debounce(setGuppyContentFromInput, 100)
).on('blur', function() {
  ...
});

```

Oishikatta replied:
Done.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+              '\\': 'shift+\\',
+              ' ': 'space',
+              ')': 'shift+0'
+              // Missing up/down. user needs to use [space], 'sub' instead.
```
What do you mean re up/down --> space/sub?

Oishikatta replied:
In the standard Guppy, the user could for example press the down arrow to go from entering an exponent to a base level variable/number, or from base to a subscript. Since those are not printable characters, this method does not proxy (replay) them to Guppy.

On a desktop browser, if you type "frac" you can then use the arrow keys to position the cursor in the top or bottom box of the fraction. On mobile, you would need to type "fraca b" to get a fraction with a as the numerator and b the denominator. *This works the same as it would on desktop, it's only that desktop users are more likely to use arrow keys than spaces.

Mobile devices typically do not have arrow key input anyways though - but still this is an area where it is likely to be unintuitive for users. How would a typical user find out that they need to press [space] to get out of things like exponents and fractions, and write out 'sub' to start a subscript? (Among other functions listed [here.](https://daniel3735928559.github.io/guppy/))

seanlip replied:
Ah! That is a _really_ good question. I don't know the answer. We can still merge this PR and file two issues to handle the question you ask in your last paragraph in a separate pass.

IMO, or the first question, I think the most intuitive thing would be to touch the relevant area on the screen. I don't know if Guppy has support for receiving such events, but it would be nice for it to have them, I think.

For the second question, I have no idea. I'm not _too_ worried, since I don't think subscripts are going to hit us for a while, but I don't know what the correct behavior should be.

In any case, could you also please file an issue (or two) [here](https://github.com/daniel3735928559/guppy/), and ask? Thank you for bringing this up!

seanlip replied:
Though touching the area on the screen is going to mess up your setGuppyContentFromInput() a bit, I suspect. You might have to implement a keystroke buffer instead that's flushed before a touch event is handled.

Oishikatta replied:
I'm not sure that issue is a good fit for the upstream repository, since it's specific to my implementation of handling mobile input. The other is also addressed upstream by having usage information on the page where Guppy is used.

I think it might be best put into a new issue (or rename #2990) here "Improve mobile usability of MathInputExpression."

My implementation of this is somewhat basic, so it doesn't currently allow any touching or selection of Guppy content.

seanlip replied:
Noted, thanks. Can you file a new issue for this, explaining clearly what is required? I agree with you that this PR can go in in the meantime.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,125 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+           * This may not be a significant issue as the MathExpressionInput is
+           * recreated if the given answer is incorrect.
+           */
+          var addMobileButtonOverlay = function() {
```
Since this is the catch-all "handle mobile" function, maybe call this makeGuppyMobileFriendly().

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -4569,3 +4569,22 @@ md-card.preview-conversation-skin-supplemental-card {
+  /* In some versions of Chrome on Android, typing in a 0px input
+  field can cause the previous character to be selected.
+  Set to 1px and 0 opacity instead of 0px to prevent this issue. */
+  width: 1px;
```
Please alphabetize.

Oishikatta replied:
Done.

seanlip commented on core/templates/dev/head/css/oppia.css:
```diff
@@ -4569,3 +4569,22 @@ md-card.preview-conversation-skin-supplemental-card {
+  margin: 0;
+  border: 0;
+  margin-left: -5px;
+}
```
Add newline at EOF.

Oishikatta replied:
Done.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,127 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+           * Adds a button overlay and invisible text field used to bring up
+           * the keyboard on mobile devices.
+           *
+           * TODO: On iOS/Safari, keyboard may only appear on first press.
```
Format should be `TODO(Oishikatta):`, so that people know who to ask for clarification.

Oishikatta replied:
Done.

Oishikatta commented:
@seanlip 

There's just two comments left, I'm unsure on them (both documentation/comments). Replied above, one still open and the other directly above it.

Aside from that it is ready. Earlier it wasn't, I hadn't finished addressing all of the comments.

seanlip commented on extensions/interactions/MathExpressionInput/MathExpressionInput.js:
```diff
@@ -30,13 +30,127 @@ oppia.directive('oppiaInteractiveMathExpressionInput', [
+              // Mobile devices typically don't have arrow keys anyways.
+            };
+
+            var setGuppyContentFromInput = function() {
```
Did you check what happens if the user hits the Backspace key -- is everything still OK? (I guess so, but just thought it worth checking.)

Oishikatta replied:
Everything is OK, if the user hits backspace the function's for loop just won't run (empty array).

Oishikatta commented:
I think this is ready to go, I looked into a couple other options but they didn't go well. So I think is a good start and we can look at improving it more later.
